### PR TITLE
Add transforming adaptation with normalizing flows

### PR DESCRIPTION
**Pull request by aseyboldt**
Experimental new algorithm that uses a normalizing flow instead of a mass matrix.

Set up using pixi:
```
git clone https://github.com/pymc-devs/nutpie
cd nutpie
git fetch origin pull/154/head:transform
git switch transform

pixi run develop
pixi shell
```
Gives a shell with an appropriate python setup.

Usage with pymc:

```python
import pymc as pm
import nutpie
import numpy as np
import jax

jax.config.update("jax_enable_x64", True)

with pm.Model() as model:
    log_sd = pm.Normal("log_sd")
    pm.Normal("y", sigma=np.exp(log_sd))

compiled = nutpie.compile_pymc_model(model, backend="jax", gradient_backend="jax")

compiled = (
    compiled
    .with_transform_adapt(
        # Neural network width, default is half the number of model parameters
        nn_width=None,
        # Number of normalizing flow layers
        num_layers=8,
        # Depth of the neural network in each flow layer
        nn_depth=1,
        # Print status update of the optimizer.
        verbose=False,
        # Number of gradients to use in each training phase
        window_size=5000,
        # Learning rate of the optimizer
        learning_rate=1e-3,
        # Print progress bars for the optimization. Very spammy...
        show_progress=False,
        # Number of initial windows with a diagonal mass matrix
        num_diag_windows=10,
    )
)

trace_ = nutpie.sample(
    compiled,
    transform_adapt=True,
    chains=2,
    tune=1000,
    draws=1000,
    cores=1,
    seed=123,
)
```

Usage with stan:

```python
import pymc as pm
import nutpie
import numpy as np
import jax
import os

os.environ["TBB_CXX_TYPE"] = "clang"
jax.config.update("jax_enable_x64", True)

code = """
parameters {
    real log_sigma;
    real x;
}
model {
    log_sigma ~ normal(0, 1);
    x ~ normal(0, exp(log_sigma));
}
"""


compiled = nutpie.compile_stan_model(code=code)

compiled = (
    compiled
    .with_transform_adapt(
        # Neural network width, default is half the number of model parameters
        nn_width=None,
        # Number of normalizing flow layers
        num_layers=8,
        # Depth of the neural network in each flow layer
        nn_depth=1,
        # Print status update of the optimizer.
        verbose=False,
        # Number of gradients to use in each training phase
        window_size=5000,
        # Learning rate of the optimizer
        learning_rate=1e-3,
        # Print progress bars for the optimization. Very spammy...
        show_progress=False,
        # Number of initial windows with a diagonal mass matrix
        num_diag_windows=10,
    )
)

trace = nutpie.sample(
    compiled,
    transform_adapt=True,
    chains=2,
    tune=1000,
    draws=1000,
    cores=1,
    seed=123,
)
```

The optimization can be quite expensive computationally (but luckily doen't need any extra gradient evaluations). A GPU is very helpful here. (Jax should pick up a cuda device automatically)

### CODE REVIEW

File: docs/nf-adapt.qmd [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+normalizing flow adaptation can also be much faster, if it can reduce the number
+of log-density evaluations needed to reach a given effective sample size.
+
+The normalizing flow adaptation works by learning a transformation of the parameter
+space that makes the posterior distribution more amenable to sampling. This is done
+by fitting a sequence of invertible transformations (the "flow") that maps the
+original parameter space to a space where the posterior is closer to a standard
+normal distribution. The flow is trained during warmup.
+
+For more information about the algorithm, see the paper todo
```
Comment (fonnesbeck) [inline docs/nf-adapt.qmd:18]: Can this link be added?

File: docs/nf-adapt.qmd [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+Currently, a lot of time is spent on compiling various parts of the normalizing
+flow, and for small models this can take a large amount of the total time.
+Hopefully, we will be able to reduce this overhead in the future.
```
Comment (fonnesbeck) [inline docs/nf-adapt.qmd:22]: Can a recommendation be made as to the size of model that warrants consideration of using NF?

File: docs/nf-adapt.qmd [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+# Define a 100-dimensional funnel model
+with pm.Model() as model:
+    log_sigma = pm.Normal("log_sigma")
+    pm.Normal("x", mu=0, sigma=pm.math.exp(log_sigma / 2), shape=100)
+
+# Compile the model with the jax backend
+compiled = nutpie.compile_pymc_model(
+    model, backend="jax", gradient_backend="jax"
+)
+```
```
Comment (fonnesbeck) [inline docs/nf-adapt.qmd:68]: Perhaps show how to use it via `pm.sample` as well?

File: docs/nf-adapt.qmd (hunk 2 above)
Comment (aseyboldt) [reply docs/nf-adapt.qmd:22]: This doesn't really depend on the size of the model.
If it is easy to sample, using a NF isn't worth it, but you can run it on small or large models.

File: docs/nf-adapt.qmd (hunk 3 above)
Comment (aseyboldt) [reply docs/nf-adapt.qmd:68]: You can't pass arguments about the normalizing flow through pm.sample right now...

Comment (aseyboldt) [reply docs/nf-adapt.qmd:68]: I added a small section in the pymc-usage notes about the `pm.sample` api.

**Comment by aseyboldt** [comment]
Closing this in favor of #180
